File: src/Frontend/App/Core/Views/MapView.cs
﻿using HikingPathFinder.Model;
using Newtonsoft.Json;
using System.Collections.Generic;
using System.Diagnostics;
using System.Globalization;
using System.Linq;
using Xamarin.Forms;

namespace HikingPathFinder.App.Views
{
    /// <summary>
    /// MapView control; the control is actually implemented as JavsScript code, but can be
    /// controlled using this class. JavaScript code is generated for function calls, and callback
    /// functions are called from JavaScript to C#.
    /// </summary>
    public class MapView
    {
        /// <summary>
        /// JSON parameters for the SetStartStopLocation event fired by JavaScript
        /// </summary>
        private class SetStartStopLocationJsonParameters
        {
            /// <summary>
            /// Indicates if start or stop location should be set
            /// </summary>
            public bool setStartLocation;

            /// <summary>
            /// Location ID of location to use
            /// </summary>
            public string locationId;

            /// <summary>
            /// Creates a new JSON parameters object
            /// </summary>
            public SetStartStopLocationJsonParameters()
            {
                this.setStartLocation = true;
                this.locationId = string.Empty;
            }
        }

        /// <summary>
        /// Web view where MapView control is used
        /// </summary>
        private readonly WebView webView;

        /// <summary>
        /// Delegate of function to call when location should be added to tour
        /// </summary>
        /// <param name="locationId">location id of location to be added</param>
        public delegate void OnAddLocationToTourCallback(string locationId);

        /// <summary>
        /// Delegate of function to call when location should be set as start or stop location
        /// </summary>
        /// <param name="setStartLocation">true when the start location should be set, false for
        /// the stop location</param>
        /// <param name="locationId">location id of location to be added</param>
        public delegate void OnSetStartStopLocationCallback(bool setStartLocation, string locationId);

        /// <summary>
        /// Delegate of function to call when navigation to location should be started
        /// </summary>
        /// <param name="locationId">location id of location to navigate to</param>
        public delegate void OnNavigateToLocationCallback(string locationId);

        /// <summary>
        /// Event that is signaled when location should be added to tour
        /// </summary>
        public event OnAddLocationToTourCallback AddLocationToTour;

        /// <summary>
        /// Event that is signaled when navigation to location should be started
        /// </summary>
        public event OnSetStartStopLocationCallback SetStartStopLocation;

        /// <summary>
        /// Event that is signaled when navigation to location should be started
        /// </summary>
        public event OnNavigateToLocationCallback NavigateToLocation;

        /// <summary>
        /// Creates a new MapView C# object
        /// </summary>
        /// <param name="webView">web view to use</param>
        public MapView(WebView webView)
        {
            this.webView = webView;

            this.webView.Navigating += this.OnNavigating_WebView;
        }

        /// <summary>
        /// Creates the MapView JavaScript object; this must be called before any other methods.
        /// </summary>
        /// <param name="areaRectangle">area rectangle to be used for map view</param>
        /// <param name="initialZoomLevel">initial zoom level, in 2D zoom level steps</param>
        public void Create(MapRectangle areaRectangle, int initialZoomLevel)
        {
            string initialRectangle = string.Format(
                "{{latitude:{0}, longitude:{1}}}, {{latitude:{2}, longitude:{3}}}",
                areaRectangle.NorthWest.Latitude.ToString(CultureInfo.InvariantCulture),
                areaRectangle.NorthWest.Longitude.ToString(CultureInfo.InvariantCulture),
                areaRectangle.SouthEast.Latitude.ToString(CultureInfo.InvariantCulture),
                areaRectangle.SouthEast.Longitude.ToString(CultureInfo.InvariantCulture));

            string js = string.Format(
                "map = new MapView({{id: 'mapElement', initialRectangle: [{0}], initialZoomLevel: {1}}});",
                initialRectangle,
                initialZoomLevel);

            this.RunJavaScript(js);
        }

        /// <summary>
        /// Zooms to given location
        /// </summary>
        /// <param name="position">position to zoom to</param>
        public void ZoomToLocation(MapPoint position)
        {
            string js = string.Format(
                "map.zoomToLocation({{latitude: {0}, longitude: {1}}});",
                position.Latitude.ToString(CultureInfo.InvariantCulture),
                position.Longitude.ToString(CultureInfo.InvariantCulture));

            this.RunJavaScript(js);
        }

        /// <summary>
        /// Updates the "my location" pin in the map
        /// </summary>
        /// <param name="position">new position to use</param>
        /// <param name="zoomToLocation">indicates if view should also zoom to the location</param>
        public void UpdateMyLocation(MapPoint position, bool zoomToLocation)
        {
            string js = string.Format(
                "map.updateMyLocation({{latitude: {0}, longitude: {1}, zoomTo: {2}}});",
                position.Latitude.ToString(CultureInfo.InvariantCulture),
                position.Longitude.ToString(CultureInfo.InvariantCulture),
                zoomToLocation ? "true" : "false");

            this.RunJavaScript(js);
        }

        /// <summary>
        /// Adds a list of locations to the map, to be displayed as pins.
        /// </summary>
        /// <param name="locationList">list of locations to add</param>
        public void AddLocationList(List<Location> locationList)
        {
            var jsonLocationList =
                from location in locationList
                select new
                {
                    id = location.Id,
                    name = location.Name,
                    isTourLocation = location.IsTourLocation,
                    type = location.Type.ToString(),
                    latitude = location.MapLocation.Latitude,
                    longitude = location.MapLocation.Longitude,
                    elevation = (int)location.Elevation
                };

            string js = string.Format(
                "map.addLocationList({0});",
                JsonConvert.SerializeObject(jsonLocationList));

            this.RunJavaScript(js);
        }

        /// <summary>
        /// Runs JavaScript code, in main thread
        /// </summary>
        /// <param name="js">javascript code snippet</param>
        private void RunJavaScript(string js)
        {
            Debug.WriteLine("run js: " + js);

            Device.BeginInvokeOnMainThread(() => this.webView.Eval(js));
        }

        /// <summary>
        /// Called when web view navigates to a new URL; used to bypass callback:// URLs.
        /// </summary>
        /// <param name="sender">sender object</param>
        /// <param name="args">event args</param>
        private void OnNavigating_WebView(object sender, WebNavigatingEventArgs args)
        {
            if (args.Url.ToString().StartsWith("callback://"))
            {
                args.Cancel = true;

                string callbackParams = args.Url.ToString().Substring(11);

                int pos = callbackParams.IndexOf('/');
                Debug.Assert(pos > 0, "callback Uri must contain a slash after the function name");

                string functionName = callbackParams.Substring(0, pos);
                string jsonParameters = callbackParams.Substring(pos + 1);

                this.ExecuteCallback(functionName, jsonParameters);
            }
        }

        /// <summary>
        /// Executes callback function
        /// </summary>
        /// <param name="functionName">function name of function to execute</param>
        /// <param name="jsonParameters">JSON formatted parameters for function</param>
        private void ExecuteCallback(string functionName, string jsonParameters)
        {
            switch (functionName)
            {
                case "onAddLocationToTour":
                    if (this.AddLocationToTour != null)
                    {
                        this.AddLocationToTour(jsonParameters.Trim('\"'));
                    }

                    break;

                case "onSetStartStopLocation":
                    if (this.SetStartStopLocation != null)
                    {
                        var parameters = JsonConvert.DeserializeObject<SetStartStopLocationJsonParameters>(jsonParameters);

                        this.SetStartStopLocation(parameters.setStartLocation, parameters.locationId);
                    }

                    break;

                case "onNavigateToLocation":
                    if (this.NavigateToLocation != null)
                    {
                        this.NavigateToLocation(jsonParameters.Trim('\"'));
                    }

                    break;

                default:
                    Debug.Assert(false, "invalid callback function name");
                    break;
            }
        }
    }
}

File: src/Frontend/App/UWP/Assets/map/mapView.js
/**
 * Creates a new instance of MapView
 * @constructor
 * @param {object} options Options to use for initializing map view
 */
function MapView(options) {

    console.log("creating new map view");

    this.options = options || {
        id: 'map',
        initialRectangle: [{ latitude: 47.77, longitude: 11.73 }, { latitude: 47.57, longitude: 12.04 }],
        initialZoomLevel: 14,
        callback: {}
    };

    if (this.options.callback === undefined)
        this.options.callback = callAction;

    var northWest = L.latLng(this.options.initialRectangle[0].latitude, this.options.initialRectangle[0].longitude);
    var southEast = L.latLng(this.options.initialRectangle[1].latitude, this.options.initialRectangle[1].longitude);
    var bounds = L.latLngBounds(northWest, southEast);
    var initialPosition = bounds.getCenter();

    this.map = L.map(this.options.id).setView(initialPosition, this.options.initialZoomLevel);

    L.tileLayer('http://{s}.tile.osm.org/{z}/{x}/{y}.png', {
        attribution: '&copy; <a href="https://osm.org/copyright">OpenStreetMap</a> contributors'
    }).addTo(this.map);

    this.myPositionMarker = L.marker([0, 0]);
    this.myPositionMarker.bindPopup('My Position');
}

/**
 * Updates the "my location" marker on the map
 * @param {object} options Options to use for updating my location. The following object can be used:
 * { latitude: 123.45678, longitude: 9.87654, zoomTo: true }
 */
MapView.prototype.updateMyLocation = function (options) {

    console.log("updating my position: lat=" + options.latitude + ", long=" + options.longitude);

    this.myPositionMarker.setLatLng([options.latitude, options.longitude]);
    this.myPositionMarker.addTo(this.map);
    this.myPositionMarker.update();

    if (options.zoomTo) {
        console.log("also zooming to my position");
        this.map.panTo([options.latitude, options.longitude]);
    }
};


/**
 * Zooms to given location
 * @param {object} options Options to use for zooming. The following object can be used:
 * { latitude: 123.45678, longitude: 9.87654 }
 */
MapView.prototype.zoomToLocation = function (options) {

    console.log("zooming to: lat=" + options.latitude + ", long=" + options.longitude);
    this.map.panTo([options.latitude, options.longitude]);
};

/**
 * Adds list of locations to the map, as marker pins
 * @param {array} locationList An array of location, each with the following object layout:
 * { id:"location-id", name:"Location Name", type:"LocationType", latitude: 123.45678, longitude: 9.87654, elevation:1234 }
 */
MapView.prototype.addLocationList = function (locationList) {

    console.log("adding location list, with " + locationList.length + " entries");

    for (var index in locationList) {

        var location = locationList[index];

        var text = '<h2>' + location.name + ' ' + location.elevation + 'm</h2>' +
            '<p>Location type: ' + location.type + '<br/>' +
            '<img height="32em" width="32em" src="images/add_to_tour.svg" style="vertical-align:middle" />' +
            '<a href="javascript:map.onAddLocationToTour(\'' + location.id + '\');">Add to tour</a> - ' +
            '<img height="32em" width="32em" src="images/route_to_location.svg" style="vertical-align:middle" />' +
            '<a href="javascript:map.onNavigateToLocation(\'' + location.id + '\');">Navigate here</a></p>';

        L.marker([location.latitude, location.longitude]).addTo(this.map).bindPopup(text);
    }
};

/**
 * Adds list of tracks to the map
 * @param {array} listOfTracks An array of tracks
 */

MapView.prototype.addTracksList = function (listOfTracks) {

    console.log("adding list of tracks, with " + listOfTracks.length + " entries");

    // TODO implement
};

/**
 * Called by the marker pin link, in order to add a location to the tour list.
 * @param {string} locationId Location ID of location to add to tour
 */
MapView.prototype.onAddLocationToTour = function (locationId) {

    console.log("location is added to tour: id=" + locationId);

    if (this.options.callback !== undefined)
        this.options.callback('onAddLocationToTour', locationId);
};

/**
 * Called by the marker pin link, in order to start navigating to the location.
 * @param {string} locationId Location ID of location to navigate to
 */
MapView.prototype.onNavigateToLocation = function (locationId) {

    console.log("navigation to location started: id=" + locationId);

    if (this.options.callback !== undefined)
        this.options.callback('onNavigateToLocation', locationId);
};
